arch/xtensa/core/gen_zsr.py

```diff
@@ -0,0 +1,61 @@
+#!/usr/bin/env python3
+# Copyright (c) 2022 Intel corporation
+# SPDX-License-Identifier: Apache-2.0
+import sys
+import re
+
+# Scratch register allocator.  Zephyr uses multiple Xtensa SRs as
+# scratch space for various special purposes.  Unfortunately the
+# configurable architecture means that not all registers will be the
+# same on every device.  This script parses a pre-cooked ("gcc -E
+# -dM") core-isa.h file for the current architecture and assigns
+# registers to usages.
+
+NEEDED = ("ALLOCA", "CPU", "FLUSH")
+
+coreisa = sys.argv[1]
+outfile = sys.argv[2]
+
+syms = {}
+
+def get(s):
+    return syms[s] if s in syms else 0
+
+with open(coreisa) as infile:
+    for line in infile.readlines():
+        m = re.match(r"^#define ([^ ]+) ?(.*)", line.rstrip())
+        if m:
+            syms[m.group(1)] = 1 if m.group(2) == "" else m.group(2)
```

Comment:
I find this quite cryptic sorry; when is `group(2)` empty and why? Could you add a couple comments or maybe even better: example(s)?

Reply:
The preprocessor behavior of "#define FOO" without a value is to expand FOO to a literal 1.  I don't know why -dM emits it like this, it seems like a wart to me.  Will comment.

Reply:
Hm... never mind.  I went back and checked, and in fact all three toolchains do this correctly and emit "#define FOO 1" and not "#define FOO".  Not sure how I got it into my head that they didn't.  Simplified.

Reply:
BTW Python regex have the useful `[\w]+` and `[\S]+` to match words, clearer and more robust than dealing with whitespace.

https://docs.python.org/3/library/re.html